modules/contact/src/actions/ContactAction.C

```diff
@@ -1028,6 +1073,166 @@ ContactAction::addNodeFaceContact()
+    if (find_set == sideset_ids.end())
+      paramError("automatic_pairing_boundaries",
+                 sideset_name,
+                 " is not defined as a sideset in the mesh.");
+
+    auto dofs_set = bnd_to_elem_map.find(_mesh->getBoundaryID(sideset_name));
+
+    // Initialize data for sideset
+    Point center_of_gravity(0, 0, 0);
+    Real accumulated_sideset_area(0);
+
+    // Pointer to lower-dimensional element on the sideset
+    const std::unordered_set<dof_id_type> & bnd_elems = dofs_set->second;
+    std::unique_ptr<const Elem> side_ptr;
+
+    for (auto elem_id : bnd_elems)
+    {
+      const Elem * elem = _mesh->elemPtr(elem_id);
+      unsigned int side = _mesh->sideWithBoundaryID(elem, _mesh->getBoundaryID(sideset_name));
+
+      // update side_ptr
+      elem->side_ptr(side_ptr, side);
+
+      // area of the (linearized) side
+      const auto side_area = side_ptr->volume();
+
+      // position of the side
+      const auto side_position = side_ptr->true_centroid();
+
+      center_of_gravity += side_position * side_area;
+      accumulated_sideset_area += side_area;
+    }
+
+    // Average each element's center of gravity with its area
+    center_of_gravity /= accumulated_sideset_area;
+
+    // Add sideset-cog pair to vector
+    automatic_pairing_boundaries_cog.emplace_back(sideset_name, center_of_gravity);
+  }
+
+  // Vectors of distances for each pair
+  std::vector<std::pair<std::pair<BoundaryName, BoundaryName>, Real>> pairs_distances;
+
+  // Assign distances to identify nearby pairs.
+  for (std::size_t i = 0; i < automatic_pairing_boundaries_cog.size() - 1; i++)
+    for (std::size_t j = i + 1; j < automatic_pairing_boundaries_cog.size(); j++)
+    {
+      const Point & distance_vector =
+          automatic_pairing_boundaries_cog[i].second - automatic_pairing_boundaries_cog[j].second;
+
+      if (automatic_pairing_boundaries_cog[i].first != automatic_pairing_boundaries_cog[j].first)
+      {
+        const Real distance = distance_vector.norm();
+        const std::pair pair = std::make_pair(automatic_pairing_boundaries_cog[i].first,
+                                              automatic_pairing_boundaries_cog[j].first);
+        pairs_distances.emplace_back(std::make_pair(pair, distance));
+      }
+    }
+
+  const auto automatic_pairing_distance = getParam<Real>("automatic_pairing_distance");
+
+  // Loop over all pairs and add those whose distance isn't trivial
+  // and less than the threshold distance
+  std::vector<std::pair<std::pair<BoundaryName, BoundaryName>, Real>> lean_pairs_distances;
+
+  // Keep track of repeated boundaries (i.e. center of gravity distance is zero)
+  std::vector<BoundaryName> repeated_boundary;
+
+  for (const auto & pair_distance : pairs_distances)
+  {
+    if (pair_distance.second > TOLERANCE && pair_distance.second <= automatic_pairing_distance)
+      lean_pairs_distances.emplace_back(pair_distance);
+    else if (pair_distance.second <= TOLERANCE)
```

Comment:
I did this per your request!

Reply:
I requested this change mostly to improve on the loop structure in the original version of this code. Then upon giving this some thought, I realized that there really are good use cases for the case where the distance is 0. I regret the inconvenience I caused by changing requests, but iterative changes in direction are inherent in software development. 

I would really prefer that we get rid of this use of tolerance (and the warning associated with it) entirely because it can give us false positives about potential issues. Legitimate uses of this are still going to trigger these warnings.

To check for this particular user error, it would be better to go through each primary/secondary pair face by face to make sure they're not identical. I guess while we're at it, we really should be more stringent and make sure that a given face does not appear in both the primary and secondary boundaries. That check should be done for all uses of the contact action, not just for this automatic pairing, because that is not an issue specific to this feature. The downsides to doing that is that it doesn't work with pre-split meshes, and it could be slow (but probably not all that slow).

I think it's a good idea to do that sometime, but in a different PR. Let's leave this so it generates a warning now, and then remove that warning later when we implement the more rigorous check.